### PR TITLE
Implement core game logic

### PR DESCRIPTION
**Pull request by BelfDev**
# Core Game Logic

- [x] Refactor all core classes
- [x] Implement a full game cycle
- [x] Introduce the observer pattern to communicate with the CLI App
- [x] Refactor the CommandLineUI to implement the GameEventListener
- [x] Enforce the MVP architectural pattern for the CLI App
<br>

**TODOS:**
- [ ] Optimize the game logic even further
- [ ] Write JavaDoc and comments
- [ ] Write the Unit tests
- [ ] Integrate core logic with the online App

### CODE REVIEW

File: src/main/java/com/toptrumps/core/card/Card.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    public Card(String name, ArrayList<Attribute> attributes) {
+        this.name = name;
+        this.attributes = attributes;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    // TODO: Revisit this method to increase efficiency
```
Comment (hughwin) [inline src/main/java/com/toptrumps/core/card/Card.java:26]: Possibly consider HashMaps! Time complexity of O(1)

Comment (BelfDev) [reply src/main/java/com/toptrumps/core/card/Card.java:26]: I sure will! :)

File: src/main/java/com/toptrumps/core/player/AIPlayer.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    private static final String DEFAULT_AI_NAME = "AI_Player";
+
+    public AIPlayer(int id) {
+        super(id, String.format("%s_%d", DEFAULT_AI_NAME, id));
+    }
+
+    public Attribute selectAttribute() {
+        Card topCard = getTopCard();
+        Attribute highestAttribute = topCard.getHighestAttribute();
```
Comment (hughwin) [inline src/main/java/com/toptrumps/core/player/AIPlayer.java]: The `getHighestAttribute()` doesn't exist in card?

**Review by hughwin — APPROVED** [review]

**Review by Arden488 — APPROVED** [review]

**Review by rjs68 — APPROVED** [review]
Oh baby

**Pull request merged**